src/mixins/createStatefulListenersMixin.ts

```diff
@@ -0,0 +1,158 @@
+import { ComposeFactory } from 'dojo-compose/compose';
+import createEvented, { Actionable, EventedListenersMap, TargettedEventObject } from 'dojo-compose/mixins/createEvented';
+import createStateful, { Stateful, StatefulOptions, StateChangeEvent } from 'dojo-compose/mixins/createStateful';
+import { Handle } from 'dojo-core/interfaces';
+import Map from 'dojo-core/Map';
+import Promise from 'dojo-core/Promise';
+import WeakMap from 'dojo-core/WeakMap';
+
+export interface ActionRegistry {
+	get(id: string | symbol): Promise<Actionable<TargettedEventObject>>;
+	identify(value: Actionable<TargettedEventObject>): string | symbol;
+}
+
+export type ListenerOrArray = string | symbol | (string | symbol)[];
+
+export interface ListenersMap {
+	[type: string]: ListenerOrArray;
+}
+
+export interface StatefulListenersState {
+	listeners?: ListenersMap;
+}
+
+export interface StatefulListenersOptions<S extends StatefulListenersState> extends StatefulOptions<S> {
+	actionRegistry?: ActionRegistry;
+}
+
+export type StatefulListeners<S extends StatefulListenersState> = Stateful<S>;
+
+export interface StatefulListenersMixinFactory extends ComposeFactory<Stateful<StatefulListenersState>, StatefulListenersOptions<StatefulListenersState>> {
+	(options?: StatefulListenersOptions<StatefulListenersState>): Stateful<StatefulListenersState>;
+}
+
+function resolveListeners(registry: ActionRegistry, cache: Map<string | symbol, Actionable<TargettedEventObject>>, ref: ListenerOrArray): { value?: any; promise?: Promise<any>; } {
```

Comment:
I guess I am "comfortable" with this, but the TypeScript convention is to use a tuple for multiple returns:

```typescript
function resolveListeners(registry: ActionRegistry, cache: Map<string | symbol, Actionable<TargettedEventObject>>, ref: ListenerOrArray): [ any, Promise<any> ]
```

I guess do we want to stick with TypeScript convention or try to do something different.  I would say unless there is a compelling reasons to break with convention...